### ISSUE TITLE
WP2 - PlusD: Data Cleaning

### ISSUE THREAD

**Issue by agulfy** [fouweric/wikileaks-data-analytics#5]
Requires a group of 2-3 people.

The group responsible for WP1 will generate a directory of html files similar to http://search.wikileaks.org/plusd/cables/1973STATE097844_b.html .

This group needs to implement a parser to extract the important bits. The output will be saved directly into the database.

**Comment by fbriant** [comment]
I'm interested in joining this group, if you didn't plan otherwise.

Should we consider the classification already done by wikileaks or the raw data to extract the important bits ?

**Comment by agulfy** [comment]
Sure, nobody is assigned to this task yet.

> Should we consider the classification already done by wikileaks or the raw data to extract the important bits ?
I didn't quite I understand this

**Comment by fbriant** [comment]
Yeah sorry, I'm a bit clumsy with english.

First, what do you consider to be the "important bits" ?

It seems, in the HTML page you gave us, that wikileaks has already extracted some "tags" from the raw text (All is in the header box; date, from, to, classification and so on). Should we use this or compute the same results from the original text ?
I realise that my question is a bit dumb, because of course it would be better to use what was already done, but I just want to be clear with what you expect from this task ;-)

**Comment by fbriant** [comment]
And another question ;

If I understood correctly, these HTML files will be on the cluster right ?
In this case, I think we should probably implement the parser in Java and use the Hadoop Java API (which we had the opportunity to work with thanks to Homework 1).
Do you think this is appropriate ?

**Comment by fouweric** [comment]
I join you in this task @fbriant .
We can mockup 10 HTML documents on hdfs. (?)
Yes, Hadoop Java API is fine!

**Comment by fbriant** [comment]
Should we use any API for HTML parsing or implement our own algorithm ?

It seems pretty easy as everything we want (if I understood the task correctly) is in 
`<table id="synopsis">`

And should we extract the text of the document as well ?

**Comment by agulfy** [comment]
I think I will join this task as well.

> And should we extract the text of the document as well ?

Let's also get
    <div id="pane-contents"> </div>

for the raw content. And finally
    <table class="references">

for the references.

**Comment by fbriant** [comment]
May we use external library ?
I found this interesting open source Java library for HTML Parsing : http://jsoup.org/

**Comment by christophkoch** [comment]
You can use all external libs that you find useful.

**Comment by fbriant** [comment]
Great thank you. It is indeed simpler with this.
Should we mention anywhere other than the source code that we used it ? Some kind of "libograpy" ?

**Comment by ElSeidy** [comment]
Yes please keep track of all the libraries that you used, which will be handed in the final report. Later on, you will probably also need to use some NLP algorithms, as mentioned to the group leaders, I would suggest using the Stanford packages 
http://nlp.stanford.edu/software/corenlp.shtml

**Comment by christophkoch** [comment]
Sure, that wouldn't hurt. In any case, the goal is to get the job done. No need to implement stuff you can legally (and freely) find out there.

**Comment by fbriant** [comment]
As stated by the last commit, the HTML parsing is ready. However, it is for a local computer, I must modify it for the cluster and a mapreduce task.

**Comment by fouweric** [comment]
Data is ready on hdfs:///projects/wikileaks/plusd/ (and issue #1 is closed)
You might also depend on issue #9 (not yet done)?

**Comment by agulfy** [comment]
> Data is ready on hdfs:///projects/wikileaks/plusd/ (and issue #1 is closed)

Great

> You might also depend on issue #9 (not yet done)?

Yes, the output from the parser (sql insert commands?) depends on #9 .

**Comment by fbriant** [comment]
Okay !
It will be done by the end of this week.
But indeed, I do need the schema of #9 for my output.

**Comment by agulfy** [comment]
Hey, I did some stuff in 92648dcd8b8dc03c3b01e9486fd23ab669fc007d. @fbriant @xPh3n0m How is the input file stuff?

**Comment by fouweric** [comment]
@agulfy Shouldn't you do sql bulk INSERT? http://stackoverflow.com/questions/12365992/sql-bulk-insert-statement

It is really faster for importing data into database :))

**Comment by agulfy** [comment]
Makes sense!

I may possibly need to change the mapper/reducer for this so I will still wait for the input mechanism to be finished.

**Comment by ElSeidy** [comment]
Are you guys using SparkSQL? or is this a single instance DB?

**Comment by agulfy** [comment]
@fbriant if you can modify the inputfileformat such that the mapper input VALUE is a Text with the full contents of one (unmerged) html file, this task is finished :) (The key is irrelevant and can be NullWritable but I kept it as Text for now)

**Comment by fbriant** [comment]
Yeah I'm going to do this right now ;-)

**Comment by fbriant** [comment]
Ok, I tried something.

However, I'm not sure the JSOUP library can be used for customised HTML tags (such as nb_documents, document_id and document_content).

As I have some problem right now with VPN connection with EPFL, may I ask you @agulfy to try my code ? Sorry for this inconvenience.

I followed this format :

<nb_documents> N </nb_documents>
<document_id id=DOCNAME1 />
<document_id id=DOCNAME2 />
<document_id id=DOCNAME3 />
...
<document_id id=DOCNAMEN />

<document_content id=DOCNAME1>
content of doc 1
</document_content>
<document_content id=DOCNAME2>
content of doc 2
</document_content>
.....

**Comment by agulfy** [comment]
I got 
```
java.lang.Exception: java.lang.NullPointerException
	at org.apache.hadoop.mapred.LocalJobRunner$Job.runTasks(LocalJobRunner.java:462)
	at org.apache.hadoop.mapred.LocalJobRunner$Job.run(LocalJobRunner.java:522)
Caused by: java.lang.NullPointerException
	at datacleaning.HTMLParsingRecordReader.initialize(HTMLParsingRecordReader.java:70)
```

Looks like `doc.select("nb_documents")` is empty so I think JSOUP can't handle custom tags or needs some configuration.

**Comment by fbriant** [comment]
Ok thanks for the information, I will take care of this !

**Comment by fbriant** [comment]
So apparently, we can configure JSOUP to work without HTML specs and with XML format, thus working with custom tags.

Could you test once more @agulfy ?
Thanks for the help !

**Comment by agulfy** [comment]
Hey, sorry I just used a wrong input. Ignore my previous comment - I will try again in an hour and let you know

**Comment by fbriant** [comment]
Thank you ! Either way, I think you were right ;-)
By googling a bit, I found we can not parse custom tags like this with HTML specs ;-)

**Comment by agulfy** [comment]
Hey @fbriant I just managed to finish the job with a few minor changes at 7b064e5f24fc4e38c7e26febd8a1602b3606e59f

But, processing one merged file (~1500-2000 cables) took around 6 minutes and we have 1050 files to process... I will see if I can increase the performance before feeding everything in to the program.

**Comment by agulfy** [comment]
Done. You can find the data on the project folder in the cluster and some examples in the repo.
